File: bots/intgrah/v54.7.2/builder/tasks/shared/opportunistic_attack.py
"""Cheap, low-priority opportunistic fire used by ECON / DEFENSE roles.
A small fraction of builders (`self.opportunistic` set at init) randomly
fire (p=0.2) on the enemy building under their feet, but only after round
100. Distinct from OFFENSE's structured attack cascade — this is just
"if standing on an enemy thing, occasionally hit it.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, override

from builder.tasks.rejected import Reason, TaskRejectedError

if TYPE_CHECKING:
    from cambc import Controller

    from builder import Builder


class TaskRejectedNotOpportunisticError(TaskRejectedError):
    @override
    def reason(self) -> Reason:
        return "builder is not in opportunistic mode"


class TaskRejectedRngDeclinedError(TaskRejectedError):
    @override
    def reason(self) -> Reason:
        return "random gate (p=0.2) declined"


class TaskRejectedTooEarlyError(TaskRejectedError):
    def __init__(self, round_: int) -> None:
        self.round = round_

    @override
    def reason(self) -> Reason:
        return "round {round} <= 100", {"round": self.round}


class TaskRejectedCannotFireError(TaskRejectedError):
    @override
    def reason(self) -> Reason:
        return "ct.can_fire(my_pos) is False"


class TaskRejectedFriendlyTileError(TaskRejectedError):
    @override
    def reason(self) -> Reason:
        return "tile under builder holds a friendly building"


def opportunistic_attack(self: Builder, ct: Controller) -> None:
    if not self.opportunistic:
        raise TaskRejectedNotOpportunisticError
    if self.rng.random() >= 0.2:
        raise TaskRejectedRngDeclinedError
    if self.round <= 100:
        raise TaskRejectedTooEarlyError(self.round)
    if not ct.can_fire(self.my_pos):
        raise TaskRejectedCannotFireError
    if ct.get_team(ct.get_tile_building_id(self.my_pos)) == self.my_team:
        raise TaskRejectedFriendlyTileError
    ct.fire(self.my_pos)

File: bots/rs/v56/src/builder/tasks/shared/opportunistic_attack.rs
//! Translation of `bots/intgrah/v54.7.9/builder/tasks/shared/opportunistic_attack.py`.
//!
//! Cheap, low-priority opportunistic fire used by ECON / DEFENSE roles.
//! A small fraction of builders (`self.opportunistic` set at init) randomly
//! fire (p=0.2) on the enemy building under their feet, but only after round
//! 100. Distinct from OFFENSE's structured attack cascade — this is just
//! "if standing on an enemy thing, occasionally hit it".

use cambc::{Controller, ControllerApi};

use crate::builder::Builder;
use crate::builder::tasks::rejected::{TaskRejected, TaskResult};

pub fn opportunistic_attack(self_: &mut Builder, ct: &mut Controller<'_>) -> TaskResult {
    if !self_.opportunistic {
        return Some(TaskRejected::new("builder is not in opportunistic mode"));
    }
    let r = self_.rng.random();
    if r >= 0.2 {
        return Some(TaskRejected::new("random gate (p=0.2) declined"));
    }
    if self_.round <= 100 {
        return Some(TaskRejected::from_string(format!(
            "round {} <= 100",
            self_.round
        )));
    }
    if !pyrust::unwrap!(ct.can_fire(self_.my_pos)) {
        return Some(TaskRejected::new("ct.can_fire(my_pos) is False"));
    }
    let bid = pyrust::unwrap!(ct.get_tile_building_id(self_.my_pos));
    if pyrust::unwrap!(ct.get_team(bid)) == self_.my_team {
        return Some(TaskRejected::new(
            "tile under builder holds a friendly building",
        ));
    }
    pyrust::unwrap!(ct.fire(self_.my_pos));
    None
}
